### PR TITLE
Optimized SpriteFont GlyphPacker

### LINKED ISSUE #7351: LocalizedFontProcessor extremely poor performance
LocalizedFontProcessor tends to have a very large number of glyphs, and its performance tanks terribly as the number increases. Actually I believe what needs to be optimized is the glyph arrangement algorithm in the base type FontDescriptionProcessor, which currently has performance like or worse than N^N. 

#### What version of MonoGame does the bug occur on:
- Console branch

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

-

### PR DESCRIPTION
**Pull request by TechnologicalPizza**
Fixes #7351

### CODE REVIEW

**Comment by TechnologicalPizza** [comment]
Sorry for the metric ton of formatting changes but couldn't bother undoing it after the IDE did it 😅

**Comment by jamesford42** [comment]
> Sorry for the metric ton of formatting changes but couldn't bother undoing it after the IDE did it 😅

This is likely to be an issue with getting it merged... while the previous formatting was certainty a bit bizarre (1-space indents??), ideally the formatting would be the first commit by itself.

**Comment by TechnologicalPizza** [comment]
The commit with formatting changes actually replaced tabs with spaces so I think that's fine.

**Comment by mrhelmut** [comment]
> The commit with formatting changes actually replaced tabs with spaces so I think that's fine.

It makes the PR hard to read, but likely is for the better good in the future since spaces are the suggested formatting on the repo'.

**Comment by TechnologicalPizza** [comment]
>maybe we should clean the formatting on the entire repo in a separate PR? 

This went from 0 to 100 real quick. 
Such PR would change a massive amount of files as the formatting is quite funny throughout the codebase.

**Comment by Jjagg** [comment]
I've talked with @harry-cpp about enforcing formatting guidelines and applying them in a big PR before. 

Combining everything in one commit reduces the impact on history, so that's a lot better than having formatting changes in random PRs that don't intend to touch formatting. We'll get to it someday 😬

Don't worry about the formatting changes here.

**Comment by harry-cpp** [comment]
For us reviewers we can always add a `?w=1` to ignore whitespace changes, ie: https://github.com/MonoGame/MonoGame/pull/7464/files?w=1

**Comment by Jjagg** [comment]
GitHub has had UI for that for a while now:
image

File: MonoGame.Framework.Content.Pipeline/Graphics/Font/GlyphPacker.cs [hunk 1]
```diff
@@ -7,203 +7,145 @@
...
+            // Work out how big the output bitmap should be.
+            int guessedWidth = GuessOutputWidth(sourceGlyphs);
+            var rectPacker = new MaxRectsBin(guessedWidth, 1024 * 16, GrowRule.Height);
 
-				outputHeight = Math.Max(outputHeight, glyphs[i].Y + glyphs[i].Height);
-			}
+            for (int i = 0; i < glyphs.Count; i++)
+            {
+                ArrangedGlyph glyph = glyphs[i];
+                Rectangle bounds = rectPacker.Insert(glyph.Width, glyph.Height, MaxRectsHeuristic.Bl);
```
Comment (stromkos) [inline MonoGame.Framework.Content.Pipeline/Graphics/Font/GlyphPacker.cs:41]: MaxRectsHeuristic.Bl, the "tertis" heuristic, can rotate the rectangle to fit.  This heuristic has the opportunity to lead to skewed fonts.

Comment (TechnologicalPizza) [reply MonoGame.Framework.Content.Pipeline/Graphics/Font/GlyphPacker.cs:41]: Rotation is disabled when creating MaxRectsBin (5 lines above; parameter `allowRotation` is false by default)

**Comment by PhantomBadger** [comment]
Didn't spot this prior to making my own PR #7637, not sure if any of the minimising-search methods I implemented will help here (or whether your changes make them redundant) but may be worth a pass.

**Comment by TechnologicalPizza** [comment]
This PR throws the old code out the window and uses a quadtree instead, with truly impressive performance at that. You should clone and try your 65k characters with it, because I managed to build some massive sprites in mere seconds.

**Comment by tomspilman** [comment]
Yeah please give it a try if you can @PhantomBadger.

**Comment by stromkos** [comment]
> Didn't spot this prior to making my own PR #7637, not sure if any of the minimising-search methods I implemented will help here (or whether your changes make them redundant) but may be worth a pass.

@PhantomBadger, I wrote my [own version of this](https://github.com/MonoGame/MonoGame/issues/7351#issuecomment-797106549) that runs in: O(n log n), for the sort + O(2n - ~log n), time in n space, but it was not better than this PR for most cases.

**Review by stromkos — APPROVED** [review]

**Comment by TechnologicalPizza** [comment]
Built a sheet with allegedly 63454 glyphs of MS Gothic at 12px. Took ~5 seconds.
image
image

**Comment by PhantomBadger** [comment]
Awesome results! Thank you for sharing :) I'll grab this fork when I next need to rebuild my font.

**Comment by mrhelmut** [comment]
I've been meaning to review this for quite some time. Hopefully it'll be my next priority.

**Comment by mrhelmut** [comment]
I actually took some time just now, it's looking good, thanks!

**Comment by juliand665** [comment]
This also seems to fix the horrible kerning that we've been experiencing:
unknown

Unfortunately, we can't easily use it because nightly builds broke before it was merged :(

**Comment by TechnologicalPizza** [comment]
I am not really sure how kerning could have been magically fixed. It may have been that the old layout algorithm misplaced glyphs in the bitmap.

**Comment by mrhelmut** [comment]
I don't see anything that would impact kerning in this PR but who knows.

**Comment by harry-cpp** [comment]
You can use nightly builds directly from GitHub packages: https://github.com/MonoGame/MonoGame/packages

I'll be posting instructions for it and a proper announcement later this week.

**Comment by juliand665** [comment]
Ah yeah, my bad: after some manual bisection it seems this wasn't what fixed it. It also works when running from commit 1970d031b3c8508bf21732742e840ac97a073fe8.

I can't (easily) check earlier commits because they don't build out of the box on my mac. Using the newest TeamCity build, 3.8.1.2097, didn't fix it, so either something happened between de90766423ec9785410ff5223bcbd5c2e757a015 and 1970d031b3c8508bf21732742e840ac97a073fe8 or the mere act of using MonoGame from source fixes it? The bug is in MGCB, so perhaps our setup is somehow stuck on the old MGCB version despite using a newer MonoGame build?

**Comment by stromkos** [comment]
@mrhelmut ,
nkast fixed the kerning issues with PR #7501

**Comment by juliand665** [comment]
That's weird, because that should have been in version 3.8.1.2097, no? We were using that and it was still broken, like I said.

**Comment by stromkos** [comment]
> That's weird, because that should have been in version 3.8.1.2097, no? We were using that and it was still broken, like I said.

While I cannot speak for build numbers, since all of my builds are from develop, my need for manual font placement adjustments(baked into the spritefont XNB file) have been minimized with this PR update, my last update was long before #7501.

I only update between projects.

**Comment by juliand665** [comment]
Ah, I should have updated this thread: I ended up figuring out why it worked from source despite seeming to have broken after 1614. When building from source, I was unknowingly falling back on the content processor from 1614 that I had previously installed. The .fx code it produced would have been incompatible with the newer engine, but I was using a setup to remotely run MGFXC on a Windows computer (since I'm on a mac), which meant I was still compiling effects with the new version while processing all other content with the old one.

And indeed, the kerning looked much better in 1614. There are some minor places where the new version does better, but overall it's visibly less regular, as seen in my examples earlier. I'm not sure where exactly it broke, but I'm not really using MonoGame anymore so I can't really look into this further.

**Comment by stromkos** [comment]
I am glad you figured it out. Thank you for letting me know.

> I'm not really using MonoGame anymore so I can't really look into this further.

I am sorry to hear that.  With regards to build sizes and code efficiency, you can't beat Monogame without writing your own customized C++ engine.

I wish you luck in all of you endeavors.

**Pull request merged**